### ISSUE TITLE
ucx-py deadlock

### ISSUE THREAD

**Issue by quasiben** [rapidsai/ucx-py#140]
In dask we have code which looks like the following:

```python
class UCX:
    def __init__(self, ep):
        self.ep = ep
    def async read()
    def async write()

class UCXListener:
    def __init__(self):

    def start(self):
        async def serve_forever(ep, li):
            ucx = UCX(ep)
            self.comm = ucx

        ucp.init()
        loop = asyncio.get_event_loop()

        self.ucp_server = ucp.start_listener(serve_forever,
                                        listener_port=13337,
                                        is_coroutine=True)
        t = loop.create_task(self.ucp_server.coroutine)
        self._t = t
```
https://github.com/dask/distributed/blob/4d98bb5d6e8b45270c81e45a5662f1b1d4edee24/distributed/comm/ucx.py#L91-L150

The above is infrastructure for defining a read/write scheme and an endpoint/listener object for ucx.  We can use the objects like so

```python
uu = UCXListener()
ucx = UCX(uu.client)
ucx.start()
await ucx.write()
await uu.comm.read()
```

@Akshay-Venkatesh pointed out that for larger messages UCX will invoke a [Rendezvous-Protocol](https://github.com/openucx/ucx/wiki/Rendezvous-Protocol).   During `Rendezvous-Protocol` the sender and receiver must go through a handshake before transmission of messages can proceed.  I believe that during this message passing protocol, UCX is using UDP style messaging -- that is, write messages to a scratch buffer, send nonblocking, and return.  By default this message size, where Rendezvous is invoked, is set to auto (we can control this with a env var if desired) and for a DGX machine this probably starts around 16KB/32KB. 

In the above, we `await write()` -- this can't proceed until there is a handshake confirmation.  We also, `await read` and this can't proceed until the write is finished.  And thus we have a deadlock.  We can do one of two things:

- don't await write -- I forget why this works now?  maybe we need two separate coroutines main/read and by not doing `await write` the handshake is established
- move read logic inside of `start_listener` coroutine.

The dask code base heavily relies on `await read/await write`: https://github.com/dask/distributed/blob/master/distributed/core.py#L525  .  I think we can proceed by putting the read logic inside an internal function and call that in `start_listener`.  When we get a response, put that in a queue, and consume the queue in a the public read()

PR https://github.com/rapidsai/ucx-py/pull/139 attempts to do this

**Comment by Akshay-Venkatesh** [comment]
@quasiben I should point out that the deadlock case shows up for the test because the same process is trying to write and read (or in message-passing terms -- send to self and receive from self). 

For these sends and receives, between 0 bytes to 16KB/32KB, eager protocol is used which means that send doesn't wait for the receiver to post a receive. The sender simply copies the source buffer to a bounce buffer, issues a non-blocking network operation and completes. The receiver awaits for the network operation to complete and be received into the destination buffer. 

Beyond 16KB-32KB, where rndv protocol is used, the send will block (if called with `await ep.send_obj` and not when just calling `ep.send_obj`) until the receiver has posted a matching receive and the sender gets information about where to write to on the remote side. This is referred as zero-copy protocol because there are intermediate copies to bounce buffers. After the initial handshake with control information, the sender's source contents are copied to receiver's and there's another handshake between sender and receiver to notify that the buffers are ready for reuse/use.

*Since the sender and receiver is the same entity in Ben's case*, the send operation is blocked until a receive is posted but the receive can't be posted with `await ep.send_obj` blocking the send until actually completes in the case of a rndv transfer. It can only complete when the receive has a chance to be posted. For an eager transfer, `await` blocks the send until it completes, but the send actually completes as soon as there is a local copy from the original send buffer to a local bounce buffer. If you don't prefix the `ep.send_obj` with `await` then a request is returned and the send is not blocking. In which case, irrespective of the protocol, you allow the send and the receive to be posted and hence you avoid the deadlock. The other alternative is to make both send and receive non-blocking, and await on the requests returned by them only when the send or receive buffer is actually needed in the program. 

Can this happen when the sending/receiving entities are different? Yes. The following is an example of a two-process case which tries to exchange message of size > rendezvous threshold and deadlocks:
```
if proc1:
   await ep2.send_obj(...)
   await ep2.recv_obj(...)
else: # if proc2
   await ep1.send_obj(...)
   await ep1.recv_obj(...)
```
One way to break the deadlock:
```
if proc1:
   ep2.send_obj(...)
   await ep2.recv_obj(...)
else: # if proc2
   ep1.send_obj(...)
   await ep1.recv_obj(...)
```
Another way to break the deadlock:
```
if proc1:
   sreq = ep2.send_obj(...)
   rreq = ep2.recv_obj(...)
else: # if proc2
   sreq = ep1.send_obj(...)
   rreq = ep1.recv_obj(...)

await sreq
await rreq
```

Secion 3.5 (semantics of point-to-point operations) in MPI standard goes into the details of this, if you need more information (https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report.pdf)

**Comment by quasiben** [comment]
My apologies to everyone (especially @Akshay-Venkatesh ) I think I sent us down the wrong path.  The custom_ep test I wrote was not representative of what happens inside of dask.  Instead, dask puts read/write in separate coroutines which does not result in any deadlock. 

In commit: https://github.com/rapidsai/ucx-py/pull/139/commits/caa569fe4f2806e0c92a890eaf537bd37dbfdd24 , I have updated the test and things work without issue.  So now we do:

> asyncio.gather(read(), write())
